File: aoc2017/day15/generators.py

import sys
from itertools import islice, izip

def generate(start, factor):
    while True:
        start = (start * factor) % 2147483647
        yield start

def solve1(startA, startB, n):
    genA = generate(startA, 16807)
    genB = generate(startB, 48271)

    return sum(
        1
        for (a, b) in islice(izip(genA, genB), n)
        if (a & 65535) == (b & 65535)
    )

def solve2(startA, startB, n):
    genA = (a for a in generate(startA, 16807) if a % 4 == 0)
    genB = (b for b in generate(startB, 48271) if b % 8 == 0)

    return sum(
        1
        for (a, b) in islice(izip(genA, genB), n)
        if (a & 65535) == (b & 65535)
    )

def main():
    a, b = sys.stdin.read().splitlines()
    startA = int(a[24:])
    startB = int(b[24:])

    print(solve1(
        startA=startA,
        startB=startB,
        n=40000000
    ))

    print(solve2(
        startA=startA,
        startB=startB,
        n=5000000
    ))

if __name__ == '__main__':
    main()

File: aoc2017/day15/generators.hs
#! /usr/bin/env stack
-- stack --resolver lts-9.14 --install-ghc runghc

import Data.Bits

generate :: Int -> Int -> [Int]
generate factor = iterate (\v -> (v * factor) `mod` 2147483647)

match :: (Int, Int) -> Bool
match (a, b) = (a .&. 65535) == (b .&. 65535)

solve1 :: String -> Int
solve1 input =
  let [startA, startB] = map getStart $ lines input
      genA = generate 16807 startA
      genB = generate 48271 startB
   in length . filter match . take 40000000 . zip genA $ genB

solve2 :: String -> Int
solve2 input =
  let [startA, startB] = map getStart $ lines input
      genA = filter (\v -> v `mod` 4 == 0) $ generate 16807 startA
      genB = filter (\v -> v `mod` 8 == 0) $ generate 48271 startB
   in length . filter match . take 5000000 . zip genA $ genB

getStart :: String -> Int
getStart line = read $ drop 24 line

main :: IO ()
main = interact (show . solve2)
